### PR TITLE
Brian/rebalancing token manager

### PR DESCRIPTION
**Pull request by bweick**

### CODE REVIEW

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+import { RebalancingHelperLibrary } from "../../core/lib/RebalancingHelperLibrary.sol";
+
+
+/**
+ * @title LinearAuctionPriceCurve
+ * @author Set Protocol
+ *
+ * Contract used to manage a Rebalancing Set Token
+ */
+contract RebalancingTokenManager {
```
Comment (asoong) [inline contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: As a supplementary contract, lets give this a more precise name, it should denote the assets BTCETHRebalancingManager?

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+    }
+
+    /* ============ External ============ */
+
+    /*
+     * When allowed on RebalancingSetToken, anyone can call for a new rebalance proposal
+     *
+     * @param  _rebalancingSetTokenAddress     The address of Rebalancing Set Token to propose new allocation
+     */
+    function proposeNewRebalance(
```
Comment (asoong) [inline contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: `propose`? makes sense to me that this contract should conform to the existing interface

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+
+        // Get naturalUnit and units of currentSet
+        uint256 currentSetNaturalUnit = currentSetTokenInterface.naturalUnit();
+        uint256[] memory currentSetUnits = currentSetTokenInterface.getUnits();
+
+        // Calculate wbtc dollar value in currentSet (in cents)
+        uint256 btcDollarAmount = calculateTokenAllocationAmountUSD(
+            _btcPrice,
+            currentSetNaturalUnit,
+            currentSetUnits[0],
```
Comment (asoong) [inline contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: This means btc always has to be the first component/unit stored?

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+        components[1] = ethAddress;
+
+        // Create the nextSetToken contract that collateralized the Rebalancing Set Token once rebalance
+        // is finished
+        address nextSetAddress = coreInterface.create(
+            setTokenFactory,
+            components,
+            units,
+            naturalUnit,
+            bytes32("btceth"),
```
Comment (asoong) [inline contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: Let's name these something better

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+            bytes32("btceth"),
```
Comment (asoong) [inline contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: BTCETH?

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+        uint256[] _units
+    )
+        private
+        returns (uint256)
+    {
+        // Calculate btcDollarAmount of one Set Token (in cents) 
+        uint256 btcDollarAmount = calculateTokenAllocationAmountUSD(
+            _btcPrice,
+            _naturalUnit,
+            _units[0],
```
Comment (asoong) [inline contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: Let's only reference these once and make sure they're consistent, e.g. passing in something more precise like `_bitcoinUnits` into the method. Fetch that once in `calculateNextSetUnits`

File: migrations/5_supplementary.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  if (network == "development" || network == "coverage") {
+    console.log("Exiting - Network is development");
+    return;
+  }
+
+  deployer.then(() => deploySupplmentContracts(deployer, network));
+};
+
+async function deploySupplmentContracts(deployer, network) {
+  // Deploy RebalancingTokenManager
```
Comment (asoong) [inline migrations/5_supplementary.js]: I'd take it up a notch, right now only deploy in development and include a medianizer. Also include a switch statement and different constants for each environment. Can be addressed in separate PR

File: utils/wrappers/rebalancingWrapper.ts [hunk 8]
```diff
@@ -495,4 +498,131 @@ export class RebalancingWrapper {
       priceDenominator,
     };
   }
+
+  /* ============ Rebalancing Token Manager ============ */
+
+  public async deployRebalancingTokenManagerAsync(
```
Comment (asoong) [inline utils/wrappers/rebalancingWrapper.ts]: probably could be more specific here, also changing the name of the contract

File: utils/wrappers/rebalancingWrapper.ts [hunk 9]
```diff
@@ -495,4 +498,131 @@ export class RebalancingWrapper {
...
+  public getExpectedNextSetParameters(
+    btcPrice: BigNumber,
+    ethPrice: BigNumber,
+  ): any {
+    let units: BigNumber[];
+    let naturalUnit: BigNumber;
+    if (btcPrice.greaterThanOrEqualTo(ethPrice)) {
+      const ethUnits = btcPrice.mul(new BigNumber(10 ** 10)).div(ethPrice).round(0, 3);
+      units = [new BigNumber(1), ethUnits];
+      naturalUnit = new BigNumber(10 ** 10);
```
Comment (asoong) [inline utils/wrappers/rebalancingWrapper.ts]: Feels like `new BigNumber(10 ** 10)` could be a constant. `DEFAULT_SET_TOKEN_NATURAL_UNIT`

File: utils/wrappers/rebalancingWrapper.ts [hunk 10]
```diff
@@ -495,4 +498,131 @@ export class RebalancingWrapper {
...
+
+  private computeTokenValue(
+    units: BigNumber[],
+    naturalUnit: BigNumber,
+    btcPrice: BigNumber,
+    ethPrice: BigNumber,
+  ): BigNumber {
+    const FULL_TOKEN_AMOUNT = new BigNumber(10 ** 18);
+    const WBTC_TOKEN_DECIMALS = new BigNumber(10 ** 8);
+    const WETH_TOKEN_DECIMALS = new BigNumber(10 ** 18);
```
Comment (asoong) [inline utils/wrappers/rebalancingWrapper.ts]: Same, all constants here. You should maybe add these to set-protocol-utils so they can be reused.

File: contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol (hunk 3 above)
Comment (bweick) [reply contracts/supplementary/rebalancing-manager/RebalancingTokenManager.sol]: Yeah, we have to get that right on deployment, but after that it's self-reinforcing

File: contracts/supplementary/rebalancing-manager/BTCETHRebalancingManager.sol [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+            // Calculate btc units as (ethPrice/btcPrice)*100. 100 is used to add 
+            // precision. The increase in unit amounts is offset by increasing the
+            // naturalUnit by two orders of magnitude so that issuance cost is still
+            // roughly the same
+            uint256 ethBtcPrice = _ethPrice.mul(PRICE_PRECISION).div(_btcPrice);
+
+            // Create unit array and define natural unit
+            units[0] = ethBtcPrice; 
+            units[1] = PRICE_PRECISION.mul(DECIMAL_DIFF_MULTIPLIER);
+            naturalUnit = uint256(10**12);
```
Comment (felix2feng) [inline contracts/supplementary/rebalancing-manager/BTCETHRebalancingManager.sol]: Is there a way to solve this without 2 different natural units?

**Pull request merged**